### ISSUE TITLE
Agent can't find all plugins when building with Docker elastic agent 

### ISSUE THREAD

**Issue by juhasipo** [gocd/gocd#2872]
##### Issue Type

- Bug Report

##### Summary

Every now and then the build fails with error: `Did not find 'scm' plugin with id 'gerrit.cs'. Looks like plugin is missing`. I encountered this issue when testing the Docker based elastic agents with Gerrit change set and Git feature branch plugins. 

I started to investigate the issue by building my own version of the agent with more logging. I found out that the problem seems to be that the agent starts to build the job before all the plugins have been loaded. This doesn't really happen on normal agents since once they are started it may take some time before the first build job is started and then they just sit and wait for the next job with all plugins already loaded. Since the elastic agents are started just before building the job the agent has to always load the plugins before building the given job. The elastic agent doesn't have as much time to load all the plugins before actually starting the build job.

###### Basic environment details

* Go Version: 16.10.0 and 16.11.0
* JAVA Version: 1.7.0_111
* OS: Server Linux 3.13.0-32-generic, Agent Linux 4.4.0-22-generic
* Docker Elastic Agent Plugin Version: 0.5
* GoLang based GoCD agent bootstrapper Version: 0.8 (https://github.com/ketan/gocd-golang-bootstrapper/releases)

##### Steps to Reproduce

1. Setup GoCD with Docker elastic agents
2. Use one of the SCM plugins to get the material for the build (https://github.com/ashwanthkumar/gocd-build-github-pull-requests)
3. Build a job. Task can be just `echo test`.

This bug is probably reproducible with any plugin that is used on the agent but the SCM plugins are probably the best option since they are invoked at the very beginning of the build. It is easier to reproduce if job is run on many agents at the same time. For testing I used 15 agents. Maybe the bigger load on the machine running the Docker makes the bug appear more often.

##### Expected Results
Agent should always find plugins when building a job.

##### Actual Results
Sometimes build fails with error `Did not find 'scm' plugin with id 'gerrit.cs'. Looks like plugin is missing`

##### Possible Fix
Make agent wait for all the plugins plugin before starting any job or wait for a plugin to be loaded before using it.

For now, my dirty workaround is to wait the plugin to be loaded in `PluginRequestHelper#submitRequest` method before actually submitting the request to the plugin. This isn't production ready fix because I assume there a many parts of the software that use this same mechanism and I don't really know at the moment what kind of impact the fix would have.

##### Log snippets

Here is a log from one of the failed builds.

```
2016-11-16 17:10:22,090 [Thread-7 ] INFO  go.plugin.infra.FelixGoPluginOSGiFramework:108 - Starting bundleGoPluginDescriptor{id='yum', version='1', about=About{name='Yum plugin', version='2.0.1', targetGoVersion='15.2.0', description='Plugin that polls a yum repository', vendor=Vendor{name='ThoughtWorks Go Team', url='www.thoughtworks.com'}, targetOperatingSystems=[Linux]}, status=PluginStatus{state=ACTIVE, messages=[], rootCauseIfInvalid=null}, pluginJarFileLocation='/go/plugins/bundled/yum-repo-exec-poller.jar', bundleLocation=plugins_work/yum-repo-exec-poller.jar, bundledPlugin=true, bundleSymbolicName='yum', bundleClassPath='lib/go-plugin-activator.jar,.,lib/httpcore-4.4.3.jar,lib/commons-logging-1.2.jar,lib/httpclient-4.5.1.jar,lib/commons-codec-1.9.jar,lib/plugin-common-1.0.jar,lib/gson-2.5.jar,lib/commons-io-2.4.jar', bundleActivator='com.thoughtworks.go.plugin.activation.DefaultGoPluginActivator', bundle=yum [1]}
2016-11-16 17:10:22,132 [loopThread] ERROR go.plugin.access.PluginRequestHelper:32 - Did not find 'scm' plugin with id 'git.fb'. Looks like plugin is missing
2016-11-16 17:10:22,142 [loopThread] ERROR go.plugin.access.PluginRequestHelper:33 -  * Available plugins are
2016-11-16 17:10:22,154 [loopThread] ERROR go.plugin.access.PluginRequestHelper:35 -    - GoPluginDescriptor{id='yum', version='1', about=About{name='Yum plugin', version='2.0.1', targetGoVersion='15.2.0', description='Plugin that polls a yum repository', vendor=Vendor{name='ThoughtWorks Go Team', url='www.thoughtworks.com'}, targetOperatingSystems=[Linux]}, status=PluginStatus{state=ACTIVE, messages=[], rootCauseIfInvalid=null}, pluginJarFileLocation='/go/plugins/bundled/yum-repo-exec-poller.jar', bundleLocation=plugins_work/yum-repo-exec-poller.jar, bundledPlugin=true, bundleSymbolicName='yum', bundleClassPath='lib/go-plugin-activator.jar,.,lib/httpcore-4.4.3.jar,lib/commons-logging-1.2.jar,lib/httpclient-4.5.1.jar,lib/commons-codec-1.9.jar,lib/plugin-common-1.0.jar,lib/gson-2.5.jar,lib/commons-io-2.4.jar', bundleActivator='com.thoughtworks.go.plugin.activation.DefaultGoPluginActivator', bundle=yum [1]}
2016-11-16 17:10:22,161 [loopThread] ERROR thoughtworks.go.work.DefaultGoPublisher:135 - Did not find 'scm' plugin with id 'git.fb'. Looks like plugin is missing
java.lang.RuntimeException: Did not find 'scm' plugin with id 'git.fb'. Looks like plugin is missing
	at com.thoughtworks.go.plugin.access.PluginRequestHelper.submitRequest(PluginRequestHelper.java:37)
	at com.thoughtworks.go.plugin.access.scm.SCMExtension.checkout(SCMExtension.java:145)
	at com.thoughtworks.go.domain.materials.scm.PluggableSCMMaterialAgent.prepare(PluggableSCMMaterialAgent.java:51)
	at com.thoughtworks.go.remote.work.BuildWork.prepareJob(BuildWork.java:193)
	at com.thoughtworks.go.remote.work.BuildWork.build(BuildWork.java:141)
	at com.thoughtworks.go.remote.work.BuildWork.doWork(BuildWork.java:96)
	at com.thoughtworks.go.agent.JobRunner.run(JobRunner.java:60)
	at com.thoughtworks.go.agent.AgentController.retrieveWork(AgentController.java:229)
	at com.thoughtworks.go.agent.AgentController.rpcLoop(AgentController.java:183)
	at com.thoughtworks.go.agent.AgentController.loop(AgentController.java:173)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.util.MethodInvoker.invoke(MethodInvoker.java:273)
	at org.springframework.scheduling.support.MethodInvokingRunnable.run(MethodInvokingRunnable.java:65)
	at org.springframework.scheduling.timer.DelegatingTimerTask.run(DelegatingTimerTask.java:70)
	at java.util.TimerThread.mainLoop(Timer.java:555)
	at java.util.TimerThread.run(Timer.java:505)
2016-11-16 17:10:22,340 [loopThread] INFO  thoughtworks.go.work.DefaultGoPublisher:105 - Agent [8f3b10acc3c3, 172.17.0.6, 6778c5fd-57db-4878-b805-2725b94ce054] is reporting build result [Failed] to Go Server for Build [gocd-slack-task-features/57/build/2/maven-runInstance-1/24061]
2016-11-16 17:10:22,442 [loopThread] INFO  thoughtworks.go.work.DefaultGoPublisher:88 - Stopping Transmission for Build [gocd-slack-task-features/57/build/2/maven-runInstance-1/24061]
2016-11-16 17:10:22,783 [loopThread] INFO  thoughtworks.go.util.HttpService:136 - Got back 200 from server
<...some log messages removed...>
2016-11-16 17:10:26,118 [Thread-7 ] INFO  go.plugin.infra.FelixGoPluginOSGiFramework:96 - Loading plugin GoPluginDescriptor{id='git.fb', version='1', about=About{name='Git Feature Branch plugin', version='1.3.1-vincit', targetGoVersion='15.1.0', description='Provides ability to do feature branch builds for Git repository', vendor=Vendor{name='Srinivas', url='https://github.com/ashwanthkumar/gocd-build-github-pull-requests'}, targetOperatingSystems=[]}, status=PluginStatus{state=ACTIVE, messages=[], rootCauseIfInvalid=null}, pluginJarFileLocation='/go/plugins/external/git-fb-poller-1.3.1-vincit.jar', bundleLocation=plugins_work/git-fb-poller-1.3.1-vincit.jar, bundledPlugin=false, bundleSymbolicName='git.fb', bundleClassPath='lib/go-plugin-activator.jar,.,lib/jackson-core-2.2.3.jar,lib/joda-time-2.3.jar,lib/annotation-indexer-1.4.jar,lib/JavaEWAH-0.7.9.jar,lib/my-java-utils-0.0.6.jar,lib/jackson-databind-2.2.3.jar,lib/commons-lang-2.6.jar,lib/httpcore-4.1.4.jar,lib/github-api-1.68.jar,lib/commons-validator-1.4.0.jar,lib/gson-2.3.1.jar,lib/commons-exec-1.3.jar,lib/bridge-method-annotation-1.14.jar,lib/commons-codec-1.7.jar,lib/jsch-0.1.51.jar,lib/org.eclipse.jgit-4.0.1.201506240215-r.jar,lib/git-cmd-1.0.jar,lib/commons-beanutils-1.8.3.jar,lib/commons-digester-1.8.jar,lib/slf4j-api-1.7.2.jar,lib/commons-logging-1.1.1.jar,lib/jackson-annotations-2.2.3.jar,lib/commons-io-2.4.jar,lib/httpclient-4.1.3.jar', bundleActivator='com.thoughtworks.go.plugin.activation.DefaultGoPluginActivator', bundle=null}
2016-11-16 17:10:26,157 [Thread-7 ] INFO  go.plugin.infra.FelixGoPluginOSGiFramework:105 - Getting bundleGoPluginDescriptor{id='git.fb', version='1', about=About{name='Git Feature Branch plugin', version='1.3.1-vincit', targetGoVersion='15.1.0', description='Provides ability to do feature branch builds for Git repository', vendor=Vendor{name='Srinivas', url='https://github.com/ashwanthkumar/gocd-build-github-pull-requests'}, targetOperatingSystems=[]}, status=PluginStatus{state=ACTIVE, messages=[], rootCauseIfInvalid=null}, pluginJarFileLocation='/go/plugins/external/git-fb-poller-1.3.1-vincit.jar', bundleLocation=plugins_work/git-fb-poller-1.3.1-vincit.jar, bundledPlugin=false, bundleSymbolicName='git.fb', bundleClassPath='lib/go-plugin-activator.jar,.,lib/jackson-core-2.2.3.jar,lib/joda-time-2.3.jar,lib/annotation-indexer-1.4.jar,lib/JavaEWAH-0.7.9.jar,lib/my-java-utils-0.0.6.jar,lib/jackson-databind-2.2.3.jar,lib/commons-lang-2.6.jar,lib/httpcore-4.1.4.jar,lib/github-api-1.68.jar,lib/commons-validator-1.4.0.jar,lib/gson-2.3.1.jar,lib/commons-exec-1.3.jar,lib/bridge-method-annotation-1.14.jar,lib/commons-codec-1.7.jar,lib/jsch-0.1.51.jar,lib/org.eclipse.jgit-4.0.1.201506240215-r.jar,lib/git-cmd-1.0.jar,lib/commons-beanutils-1.8.3.jar,lib/commons-digester-1.8.jar,lib/slf4j-api-1.7.2.jar,lib/commons-logging-1.1.1.jar,lib/jackson-annotations-2.2.3.jar,lib/commons-io-2.4.jar,lib/httpclient-4.1.3.jar', bundleActivator='com.thoughtworks.go.plugin.activation.DefaultGoPluginActivator', bundle=null}
```

In the log you can see that the `loopThread` logs an error I added to the `PluginRequestHelper#submitRequest` method. After that you can see that some of the plugins are still being loaded in `Thread-7` most notably the missing SMC plugin `git.fb`. Also note that `Available plugins are` debug print only lists one plugin when the error occurs.

**Comment by ketan** [comment]
Thank you for reporting the issue. It appears that the plugin initialisation happens in the background while the agent boots up. This may lead to a situation where the agent begins executing work, while the plugin initialisation has not completed, leading up to the situation where you see the message `Did not find 'scm' plugin with id 'gerrit.cs'. Looks like plugin is missing`.

We have a fix and we'll try and put that fix this release.

**Comment by juhasipo** [comment]
@ketan Great, thanks!

**Comment by juhasipo** [comment]
@ketan This doesn't seem to work yet in 16.12.0. The agent still accepts work before all the plugins have been loaded.

**Comment by victor-sim** [comment]
@ketan We also have the same problem with Elastic Agent in v17.12.0. If a job assigned to a newly created elastic agent and very if the job uses external plugins as a first or second task, agent fails to load plug in on time and complaining following message and fails the job.

> Error: Did not find 'task' plugin with id '<plugin id>'. Looks like plugin is missing

**Comment by tomekgajdos** [comment]
We have the same issue. Is there any plan for releasing this fix?

**Comment by arvindsv** [comment]
@tomekgajdos What's happening with your setup is probably not this. This issue was considered "fixed" through https://github.com/gocd/gocd/pull/2873, as far as I can see. It looks like the issue was not marked closed. I'll do that now. Do you mind opening a new issue with details, please? Thank you!

**Comment by tomekgajdos** [comment]
@arvindsv: New issue https://github.com/gocd/gocd/issues/5169 created.
